File: Java/RecursionPractice/src/calculations/SumOfArrayOfIntegers.java
package calculations;

import java.util.Arrays;
import java.util.List;

public class SumOfArrayOfIntegers {

    private final List<Integer> listOfIntegers;


    private int recursivelyCalculateSum(List<Integer> diminishingListOfIntegers) {
        if (diminishingListOfIntegers.isEmpty()) {
            return 0;
        }
        return diminishingListOfIntegers.get(0)
                + this.recursivelyCalculateSum(
                        diminishingListOfIntegers.subList(1, diminishingListOfIntegers.size())
        );
    }

    public int getSum() {
        return this.recursivelyCalculateSum(this.listOfIntegers);
    }

    public SumOfArrayOfIntegers(List<Integer> listOfIntegers) {
        this.listOfIntegers = listOfIntegers;
    }
}

File: Kotlin/RecursionPractice/src/calculations/SumOfArrayOfIntegers.kt
package calculations

class SumOfArrayOfIntegers(private var intArray: IntArray) {
    private fun recursivelyCalculateSum(arr: IntArray): Int {
        if (arr.size == 1) {
            return arr[0]
        }
        val removedItem = arr[0]
        return recursivelyCalculateSum(arr.sliceArray(1..<arr.size)) + removedItem
    }

    public fun getSum(): Int {
        return recursivelyCalculateSum(intArray)
    }
}